File: src/shared/items.ts
// Catálogo de itens do TBH (ver TBHPEDIA.md > Itens & Raridades e FONTES.md).
// Liga o dado bruto de `itemData.ts` (gerado pelo datamine) a rótulos PT, categorias
// de slot e metadados de raridade — base da aba Inventário (D3) e da classificação por
// raridade (D2). Tipos/raridades podem mudar com patches; regenerar `itemData.ts`.

import { GEAR_TYPE_IDS, GRADE_IDS, ITEM_DATA, ITEM_TYPE_IDS } from './itemData'

export type ItemType = (typeof ITEM_TYPE_IDS)[number] // 'GEAR' | 'MATERIAL' | 'STAGEBOX'
export type GearTypeId = (typeof GEAR_TYPE_IDS)[number]
export type GradeId = (typeof GRADE_IDS)[number]

// Categoria de slot do gear (agrupa os tipos na matriz da aba Inventário).
export type GearCategory = 'weapon' | 'offhand' | 'armor' | 'accessory'

export const CATEGORY_LABELS: Record<GearCategory, string> = {
  weapon: 'Armas',
  offhand: 'Mão secundária',
  armor: 'Armadura',
  accessory: 'Acessórios'
}

export interface GradeMeta {
  id: GradeId
  namePt: string
  tier: number // 0 (Common) .. 9 (Cosmic)
  marketable: boolean // Legendary+ é vendável no Steam Market
  color: string // cor da raridade (usada na visualização)
}

// Ordem e nomes seguem GRADE_IDS (Common → Cosmic). Cores escolhidas para a UI.
const GRADE_META: Record<GradeId, { namePt: string; color: string }> = {
  COMMON: { namePt: 'Comum', color: '#9aa4b2' },
  UNCOMMON: { namePt: 'Incomum', color: '#4ade80' },
  RARE: { namePt: 'Raro', color: '#38bdf8' },
  LEGENDARY: { namePt: 'Lendário', color: '#f59e0b' },
  IMMORTAL: { namePt: 'Imortal', color: '#ef4444' },
  ARCANA: { namePt: 'Arcano', color: '#a855f7' },
  BEYOND: { namePt: 'Além', color: '#ec4899' },
  CELESTIAL: { namePt: 'Celestial', color: '#22d3ee' },
  DIVINE: { namePt: 'Divino', color: '#eab308' },
  COSMIC: { namePt: 'Cósmico', color: '#fb7185' }
}

// Primeira raridade vendável no Market (Legendary e acima).
export const MARKETABLE_TIER = GRADE_IDS.indexOf('LEGENDARY')

export const GRADES: GradeMeta[] = GRADE_IDS.map((id, tier) => ({
  id,
  namePt: GRADE_META[id].namePt,
  tier,
  marketable: tier >= MARKETABLE_TIER,
  color: GRADE_META[id].color
}))

export interface GearTypeMeta {
  id: GearTypeId
  namePt: string
  category: GearCategory
}

// Rótulo PT + categoria de cada tipo de gear. A ordem aqui define a ordem das linhas
// na matriz tipo × raridade (agrupada por categoria).
const GEAR_TYPE_META: Record<GearTypeId, { namePt: string; category: GearCategory }> = {
  SWORD: { namePt: 'Espada', category: 'weapon' },
  AXE: { namePt: 'Machado', category: 'weapon' },
  HATCHET: { namePt: 'Machadinha', category: 'weapon' },
  BOW: { namePt: 'Arco', category: 'weapon' },
  CROSSBOW: { namePt: 'Besta', category: 'weapon' },
  STAFF: { namePt: 'Cajado', category: 'weapon' },
  SCEPTER: { namePt: 'Cetro', category: 'weapon' },
  SHIELD: { namePt: 'Escudo', category: 'offhand' },
  ARROW: { namePt: 'Flecha', category: 'offhand' },
  BOLT: { namePt: 'Virote', category: 'offhand' },
  ORB: { namePt: 'Orbe', category: 'offhand' },
  TOME: { namePt: 'Tomo', category: 'offhand' },
  HELMET: { namePt: 'Elmo', category: 'armor' },
  ARMOR: { namePt: 'Armadura', category: 'armor' },
  GLOVES: { namePt: 'Luvas', category: 'armor' },
  BOOTS: { namePt: 'Botas', category: 'armor' },
  AMULET: { namePt: 'Amuleto', category: 'accessory' },
  EARING: { namePt: 'Brinco', category: 'accessory' },
  RING: { namePt: 'Anel', category: 'accessory' },
  BRACER: { namePt: 'Bracelete', category: 'accessory' }
}

const CATEGORY_ORDER: GearCategory[] = ['weapon', 'offhand', 'armor', 'accessory']

// Tipos de gear ordenados por categoria (ordem das linhas da matriz).
export const GEAR_TYPES: GearTypeMeta[] = GEAR_TYPE_IDS.map((id) => ({
  id,
  namePt: GEAR_TYPE_META[id]?.namePt ?? id,
  category: GEAR_TYPE_META[id]?.category ?? 'accessory'
})).sort((a, b) => {
  const c = CATEGORY_ORDER.indexOf(a.category) - CATEGORY_ORDER.indexOf(b.category)
  return c !== 0 ? c : a.namePt.localeCompare(b.namePt, 'pt-BR')
})

export interface ItemInfo {
  key: number
  type: ItemType
  gearType: GearTypeId | null
  category: GearCategory | null
  grade: GradeId | null
  gradeTier: number // -1 quando o item não tem raridade
  level: number | null
  marketable: boolean
}

const gradeByTier = (tier: number): GradeMeta | undefined => GRADES[tier]

/**
 * Classifica um ItemKey do save: tipo (gear/material/baú), tipo de gear (slot),
 * raridade e nível. Retorna null para chaves desconhecidas (catálogo desatualizado).
 */
export function classifyItem(key: number | string): ItemInfo | null {
  const datum = ITEM_DATA[String(key)]
  if (!datum) return null
  const [typeIdx, gtIdx, gradeIdx, level] = datum
  const type = ITEM_TYPE_IDS[typeIdx] ?? null
  if (!type) return null
  const gearType = gtIdx >= 0 ? (GEAR_TYPE_IDS[gtIdx] ?? null) : null
  const grade = gradeByTier(gradeIdx)
  return {
    key: Number(key),
    type,
    gearType,
    category: gearType ? (GEAR_TYPE_META[gearType]?.category ?? null) : null,
    grade: grade?.id ?? null,
    gradeTier: gradeIdx,
    level: level >= 0 ? level : null,
    marketable: gradeIdx >= MARKETABLE_TIER
  }
}

export function gradeMetaById(id: GradeId): GradeMeta | undefined {
  return GRADES.find((g) => g.id === id)
}

File: dotnet/TbhTracker.Core/Logic/Items.cs
using System.Globalization;

namespace TbhTracker.Core.Logic;

// Port de src/shared/items.ts (catalogo cru vem de itemData.json).

public sealed class GradeMeta
{
    public string Id { get; init; } = "";
    public string NamePt { get; init; } = "";
    public int Tier { get; init; }
    public bool Marketable { get; init; }
    public string Color { get; init; } = "";
}

public sealed class GearTypeMeta
{
    public string Id { get; init; } = "";
    public string NamePt { get; init; } = "";
    public string Category { get; init; } = "";
}

public sealed class ItemInfo
{
    public int Key { get; init; }
    public string Type { get; init; } = "";
    public string? GearType { get; init; }
    public string? Category { get; init; }
    public string? Grade { get; init; }
    public int GradeTier { get; init; }
    public int? Level { get; init; }
    public bool Marketable { get; init; }
}

public static class Items
{
    public static readonly Dictionary<string, string> CategoryLabels = new()
    {
        ["weapon"] = "Armas",
        ["offhand"] = "Mão secundária",
        ["armor"] = "Armadura",
        ["accessory"] = "Acessórios"
    };

    private static readonly Dictionary<string, (string NamePt, string Color)> GradeMetaMap = new()
    {
        ["COMMON"] = ("Comum", "#9aa4b2"),
        ["UNCOMMON"] = ("Incomum", "#4ade80"),
        ["RARE"] = ("Raro", "#38bdf8"),
        ["LEGENDARY"] = ("Lendário", "#f59e0b"),
        ["IMMORTAL"] = ("Imortal", "#ef4444"),
        ["ARCANA"] = ("Arcano", "#a855f7"),
        ["BEYOND"] = ("Além", "#ec4899"),
        ["CELESTIAL"] = ("Celestial", "#22d3ee"),
        ["DIVINE"] = ("Divino", "#eab308"),
        ["COSMIC"] = ("Cósmico", "#fb7185")
    };

    private static readonly Dictionary<string, (string NamePt, string Category)> GearTypeMetaMap = new()
    {
        ["SWORD"] = ("Espada", "weapon"),
        ["AXE"] = ("Machado", "weapon"),
        ["HATCHET"] = ("Machadinha", "weapon"),
        ["BOW"] = ("Arco", "weapon"),
        ["CROSSBOW"] = ("Besta", "weapon"),
        ["STAFF"] = ("Cajado", "weapon"),
        ["SCEPTER"] = ("Cetro", "weapon"),
        ["SHIELD"] = ("Escudo", "offhand"),
        ["ARROW"] = ("Flecha", "offhand"),
        ["BOLT"] = ("Virote", "offhand"),
        ["ORB"] = ("Orbe", "offhand"),
        ["TOME"] = ("Tomo", "offhand"),
        ["HELMET"] = ("Elmo", "armor"),
        ["ARMOR"] = ("Armadura", "armor"),
        ["GLOVES"] = ("Luvas", "armor"),
        ["BOOTS"] = ("Botas", "armor"),
        ["AMULET"] = ("Amuleto", "accessory"),
        ["EARING"] = ("Brinco", "accessory"),
        ["RING"] = ("Anel", "accessory"),
        ["BRACER"] = ("Bracelete", "accessory")
    };

    private static readonly string[] CategoryOrder = { "weapon", "offhand", "armor", "accessory" };
    private static readonly CultureInfo PtBr = CultureInfo.GetCultureInfo("pt-BR");

    public static int MarketableTier => Catalog.Items.GradeIds.IndexOf("LEGENDARY");

    private static List<GradeMeta>? _grades;
    public static IReadOnlyList<GradeMeta> Grades => _grades ??= BuildGrades();

    private static List<GradeMeta> BuildGrades()
    {
        var ids = Catalog.Items.GradeIds;
        var marketableTier = MarketableTier;
        var list = new List<GradeMeta>();
        for (var tier = 0; tier < ids.Count; tier++)
        {
            var id = ids[tier];
            var meta = GradeMetaMap.TryGetValue(id, out var m) ? m : (id, "#9aa4b2");
            list.Add(new GradeMeta
            {
                Id = id,
                NamePt = meta.Item1,
                Tier = tier,
                Marketable = tier >= marketableTier,
                Color = meta.Item2
            });
        }
        return list;
    }

    private static List<GearTypeMeta>? _gearTypes;
    public static IReadOnlyList<GearTypeMeta> GearTypes => _gearTypes ??= BuildGearTypes();

    private static List<GearTypeMeta> BuildGearTypes()
    {
        return Catalog.Items.GearTypeIds
            .Select(id =>
            {
                var meta = GearTypeMetaMap.TryGetValue(id, out var m) ? m : (id, "accessory");
                return new GearTypeMeta { Id = id, NamePt = meta.Item1, Category = meta.Item2 };
            })
            .OrderBy(g => Array.IndexOf(CategoryOrder, g.Category))
            .ThenBy(g => g.NamePt, StringComparer.Create(PtBr, false))
            .ToList();
    }

    public static GradeMeta? GradeByTier(int tier) =>
        tier >= 0 && tier < Grades.Count ? Grades[tier] : null;

    public static GradeMeta? GradeMetaById(string id) =>
        Grades.FirstOrDefault(g => g.Id == id);

    public static ItemInfo? ClassifyItem(string key)
    {
        if (!Catalog.Items.Items.TryGetValue(key, out var datum)) return null;
        var typeIdx = datum[0];
        var gtIdx = datum[1];
        var gradeIdx = datum[2];
        var level = datum[3];

        if (typeIdx < 0 || typeIdx >= Catalog.Items.ItemTypeIds.Count) return null;
        var type = Catalog.Items.ItemTypeIds[typeIdx];

        string? gearType = gtIdx >= 0 && gtIdx < Catalog.Items.GearTypeIds.Count
            ? Catalog.Items.GearTypeIds[gtIdx]
            : null;
        var grade = GradeByTier(gradeIdx);
        string? category = gearType != null && GearTypeMetaMap.TryGetValue(gearType, out var gm)
            ? gm.Category
            : null;

        int.TryParse(key, out var keyNum);
        return new ItemInfo
        {
            Key = keyNum,
            Type = type,
            GearType = gearType,
            Category = category,
            Grade = grade?.Id,
            GradeTier = gradeIdx,
            Level = level >= 0 ? level : null,
            Marketable = gradeIdx >= MarketableTier
        };
    }
}
